k2/csrc/determinize.h

```diff
@@ -0,0 +1,837 @@
+  and it represents an un-normalized determinized state (see intro at the top
+  of this file), or an un-normalized determinized state under construction
+  (we add to it using AcceptIncomingArc()).
+
+  After we call Normalize() on it, it is a normalized determinized-state (this
+  also outputs the weight you need for the incoming arc).
+
+  After that
+
+ */
+
+template <class TracebackState>  // TracebackState == MaxTracebackState or
+                                 // LogSumTracebackState
+class DetState {
+ public:
+  using DerivType = typename TracebackState::DerivType;
+  // DerivType == int32_t for MaxTracbackState, or
+  // pair<int32_t, float> for LogSumTracebackState.
+
+  // Constructor for the initial state of the determinized FSA
+  DetState() : seq_len(0), normalized(true) {
+    // the constructor of TracebackState that takes no args gives us what we
+    // need for the start-state.
+    elements[0] = std::make_shared<TracebackState>();
+  }
+
+  /*
+     Constructor (this is the one that's normally used).
+       @param [in] seq_len  Length of symbol sequence from its
+                           base_state (this is before normalization).
+                           Will be the seq_len of the source det_state plus
+                           one.  This seq_len may end up getting reduced
+                           when Normalize() is called (reducing seq_len
+                           implicitly advances the base_state).
+   */
+  explicit DetState(int32_t seq_len)
+      : seq_len(seq_len),
+        normalized(false) {}  // .. and forward_backward_prob undefined
+
+  /**
+     Process incoming arc to this DetState.  See documentation for
+     constructor of TracebackState (which will be MaxTracebackState or
+     LogSumTracebackState).
+         @param [in] state_id  State-id, in input FSA, into which this arc
+     enters. [Note: a DetState is a weighted subset of state-ids.]
+         @param [in] src  The preceding state (from which the arc leaves).
+                    This will be a member of the `elements` of the "parent"
+                    DetState, i.e. the DetState in whose ProcessArcs() function
+                    this DetState was created.
+         @param [in] incoming_arc_index  Arc in input FSA that enters state
+                   `state_id`.
+         @param [in] arc_weight  The weight on this arc
+   */
+  void AcceptIncomingArc(int32_t state_id,
+                         const std::shared_ptr<TracebackState> &src,
+                         int32_t incoming_arc_index, int32_t arc_weight) {
+    auto ret = elements.insert({state_id, nullptr});
+    if (!ret.second) {  // No such state existed in `elements`
+      ret.first->second = std::make_shared<TracebackState>(
+          state_id, src, incoming_arc_index, arc_weight);
+    } else {  // A state with this staste_id existed in `elements`.
+      ret.first->second->Accept(src, incoming_arc_index, arc_weight);
+    }
+  }
+
+  int32_t state_id;
```

Comment:
@danpovey , I added a property of `state_id` as it seems it is used in many places (but do definition)

Reply:
OK but it needs to be documented.  E.g.
```
  // State-id in the output FSA.  Only defined after DetStateMap::GetOutputState()
  //  is called.  A DetState that is in the queue should have this defined.
```

Reply:
Added